core/src/main/java/com/sequenceiq/cloudbreak/core/flow/ClusterBootstrapperErrorHandler.java

```diff
@@ -0,0 +1,109 @@
+import java.util.Set;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.domain.CloudPlatform;
+import com.sequenceiq.cloudbreak.domain.InstanceGroup;
+import com.sequenceiq.cloudbreak.domain.InstanceMetaData;
+import com.sequenceiq.cloudbreak.domain.Resource;
+import com.sequenceiq.cloudbreak.domain.Stack;
+import com.sequenceiq.cloudbreak.orchestrator.CloudbreakOrchestratorException;
+import com.sequenceiq.cloudbreak.orchestrator.ContainerOrchestrator;
+import com.sequenceiq.cloudbreak.orchestrator.Node;
+import com.sequenceiq.cloudbreak.repository.HostMetadataRepository;
+import com.sequenceiq.cloudbreak.repository.InstanceGroupRepository;
+import com.sequenceiq.cloudbreak.repository.InstanceMetaDataRepository;
+import com.sequenceiq.cloudbreak.repository.ResourceRepository;
+import com.sequenceiq.cloudbreak.service.stack.connector.CloudPlatformConnector;
+import com.sequenceiq.cloudbreak.service.stack.resource.ResourceBuilder;
+import com.sequenceiq.cloudbreak.service.stack.resource.ResourceBuilderInit;
+
+@Component
+public class ClusterBootstrapperErrorHandler {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClusterBootstrapperErrorHandler.class);
+
+    @Autowired
+    private ResourceRepository resourceRepository;
+
+    @Autowired
+    private InstanceMetaDataRepository instanceMetaDataRepository;
+
+    @Autowired
+    private InstanceGroupRepository instanceGroupRepository;
+
+    @Autowired
+    private HostMetadataRepository hostMetadataRepository;
+
+    @javax.annotation.Resource
+    private Map<CloudPlatform, List<ResourceBuilder>> instanceResourceBuilders;
+
+    @javax.annotation.Resource
+    private Map<CloudPlatform, ResourceBuilderInit> resourceBuilderInits;
+
+    @javax.annotation.Resource
+    private Map<CloudPlatform, CloudPlatformConnector> cloudPlatformConnectors;
+
+    public void terminateFailedNodes(ContainerOrchestrator orchestrator, Stack stack, Set<Node> nodes) throws CloudbreakOrchestratorException {
+        InstanceGroup gateway = stack.getGatewayInstanceGroup();
+        InstanceMetaData gatewayInstance = gateway.getInstanceMetaData().iterator().next();
+        List<String> allAvailableNode = orchestrator.getAllAvailableNode(gatewayInstance.getPublicIp(), nodes);
+        List<Node> missingNodes = selectMissingNodes(nodes, allAvailableNode);
+        if (missingNodes.size() > 0) {
+            LOGGER.info(String.format("Bootstrap failed on %s nodes. These nodes will be terminated.", missingNodes.size()));
+            for (Node missingNode : missingNodes) {
+                InstanceMetaData hostInStack =
+                        instanceMetaDataRepository.findByPrivateAndPublicAddress(stack.getId(), missingNode.getPrivateIp(), missingNode.getPublicIp());
+                instanceMetaDataRepository.delete(hostInStack.getId());
```

Comment:
I think it would be better to delete instance metadata from the database only after the termination is completed on the cloud platform, because if that fails for some reason we won't know anything about it.
Instance metadata has a status field that can be used to (e.g: by setting it to FAILED).